### ISSUE TITLE
jbang wrapper generated files have wrong timestamp

### ISSUE THREAD

**Issue by manikmagar** [jbangdev/jbang#300]
I upgraded jbang to 0.45.0 and tried `jbang wrapper install`. It generated jbang command files but the file timestamp was 80's.

<img width="1535" alt="image" src="https://user-images.githubusercontent.com/877286/92974321-edf5e700-f453-11ea-9322-c9ba48559cee.png">

Also, the file browser is showing same - 

<img width="731" alt="image" src="https://user-images.githubusercontent.com/877286/92974383-0ebe3c80-f454-11ea-9303-f4804be9491d.png">

Java version - 
```
openjdk version "14.0.2" 2020-07-14
OpenJDK Runtime Environment AdoptOpenJDK (build 14.0.2+12)
OpenJDK 64-Bit Server VM AdoptOpenJDK (build 14.0.2+12, mixed mode, sharing)
```

I tried in two different folders and had same behavior for generated files.

Let me know if any other information is needed.

**Comment by manikmagar** [comment]
Ran with JDK 11 just in case but had same behavior - 

<img width="1536" alt="image" src="https://user-images.githubusercontent.com/877286/92974753-e7b43a80-f454-11ea-85da-11a1aade883d.png">

**Comment by chrsblck** [comment]
@maxandersen Can I work on this?

It's probably something around the `StandardCopyOption.COPY_ATTRIBUTES` 
https://github.com/jbangdev/jbang/blob/6657150f510bb398746fb6d5e7820a5c88d55d61/src/main/java/dev/jbang/cli/Wrapper.java#L71-L74

**Comment by maxandersen** [comment]
Go ahead! Work away :)

**Comment by chrsblck** [comment]
@maxandersen Do you have any suggestions here?

For example, the [Files#copy docs](https://docs.oracle.com/javase/7/docs/api/java/nio/file/Files.html#copy(java.nio.file.Path,%20java.nio.file.Path,%20java.nio.file.CopyOption...)) for `COPY_ATTRIBUTES` state:
> Attempts to copy the file attributes associated with this file to the target file. The exact file attributes that are copied is platform and file system dependent and therefore unspecified. Minimally, the last-modified-time is copied to the target file if supported by both the source and target file store. Copying of file timestamps may result in precision loss.

Seems like we can just remove this option, possibly. The other fix would probably be making a call to update the modified-timestamp after copy. wdyt?

**Comment by maxandersen** [comment]
My understand is that if you use StandardCopyOption.COPY_ATTRIBUTES best effort is done and last modified at least should be right. 

Haven't tested it though.

**Comment by chrsblck** [comment]
Actually, I couldn't replicate locally, so I installed via `Homebrew`.

```❯ la -lT /usr/local/opt/jbang/libexec
drwxr-xr-x    - cblack  9 Oct 12:47 /usr/local/opt/jbang/libexec
drwxr-xr-x    - cblack  9 Oct 12:47 ├── bin
.rwxr-xr-x 5.4k cblack  9 Oct 12:47 │  ├── jbang
.rw-r--r-- 5.0k cblack  1 Feb  1980 │  ├── jbang.cmd
.rw-r--r-- 5.9M cblack  1 Feb  1980 │  ├── jbang.jar
.rw-r--r-- 6.6k cblack  1 Feb  1980 │  └── jbang.ps1
drwxr-xr-x    - cblack  1 Feb  1980 └── examples
.rw-r--r-- 6.2k cblack  1 Feb  1980    ├── analytics.java
...
```

It looks like this is the issue.

**Comment by chrsblck** [comment]
Just for fun I checked older releases, they seem to have this issue too

```❯ la -lh jbang-0.12.2
Permissions Size User   Date Modified Name
drwxr-xr-x     - cblack  1 Feb  1980  bin
drwxr-xr-x     - cblack  1 Feb  1980  examples
```

**Comment by maxandersen** [comment]
Oh so the brew install doesn't do proper time modifications? Weird. 

Search for .rb or look in home tap Repo.

**Comment by chrsblck** [comment]
Sorry, I wasn't clear. I checked the older releases by downloading the zip files from Github. 

For example
```~/Desktop/jbang
❯ export VERSION='0.34.0' && curl -sLO https://github.com/jbangdev/jbang/releases/download/v$VERSION/jbang-$VERSION.zip

~/Desktop/jbang
❯ unzip -q jbang-0.34.0.zip

~/Desktop/jbang
❯ la -lh jbang-0.34.0
Permissions Size User   Date Modified Name
drwxr-xr-x     - cblack  1 Feb  1980  bin
drwxr-xr-x     - cblack  1 Feb  1980  examples
```

So I guess it's Github Actions, that seems unlikely though 🤔 . Still looking...

**Comment by maxandersen** [comment]
ah - so I can see it is gradle that does this . all zips have the same exact date. Believe that is done to ensure the zips stays the same between builds.

I'm wondering if we could at least set timestamp to the last commit it is built from. That would be more correct and still stable.

**Comment by aalmiray** [comment]
Could it be related to https://docs.gradle.org/current/dsl/org.gradle.api.tasks.bundling.Zip.html#org.gradle.api.tasks.bundling.Zip:preserveFileTimestamps ?

**Comment by maxandersen** [comment]
yeah, so I actually set that in build.gradle .... to have the zip stay the same no matter how often I rebuild...wonder if can do one better...and configure the timestamp used rather than 1980ies..

**Comment by maxandersen** [comment]
I guess we would need to set the timestamp somehow on all incoming files...somehow.

**Comment by chrsblck** [comment]
Sounds good, I'll look into different options and report back (after lunch 😄 )

**Comment by chrsblck** [comment]
@maxandersen I found this https://discuss.gradle.org/t/better-alternative-for-preservefiletimestamps-false/34528.

I'm not very familiar with Gradle, we use Maven, do you have easy steps to reproduce the build issue you're solving by setting `preserveFileTimestamps = false`?

**Comment by maxandersen** [comment]
what happens is that a rebuild will create a different jar even though nothing changed. that makes things slower and mutable. I'll rather keep the 1980 "feature" than breaking reproducible build if the only issue is this "surprise". 

i'm no expert in gradle neither so would need some exploration to see how can set the timestamps properly.

**Comment by aalmiray** [comment]
Yeah, I don't think you can get around this if you want to enforce reproducible builds.

**Comment by maxandersen** [comment]
if we set the files to last commit timestamp should give the same result should it not?

**Comment by chrsblck** [comment]
I'm looking, but I don't see support for setting this timestamp.

**Comment by aalmiray** [comment]
I suppose so, as long as you can fix the timestamp to a particular value that makes sense to you.

**Comment by chrsblck** [comment]
I'm not seeing a way to do this on the gradle side. Maybe gradle has a low level api or something?

**Comment by chrsblck** [comment]
@maxandersen Sorry, I haven't had time to look into this again, but I should tonight I hope. Do you have any suggestions on how to proceed?

**Comment by maxandersen** [comment]
the thing to realize is that build.gradle is just a groovy DSL and you have access to almost all groovy features. Thus one way I can think of it it is simply add a task or afterLoad section that iterates the relevant `files` collection and do a call to `Files.setAttribute(p, "creationTime", FileTime.fromMillis(c.getTimeInMillis()));`.

The tricky part is to find the right timestamp to use - here I would go search for a gradle plugin that should be able to have already handled getting git info including timestamp for latest commit.

That would be the bruteforce way to start exploring.

**Comment by aalmiray** [comment]
https://github.com/nemerosa/versioning/ is my preferred choice for git in Gradle, sadly for for this case it may not work given that it does no give you access to the commit date as far as I can tell.

https://github.com/ajoberstar/gradle-git-publish is the other one I rely on a lot. It uses https://github.com/ajoberstar/grgit which wraps jgit, thus you may be able to get the date form it. Or simply use a Java process to invoke git underneath ;-)

**Comment by chrsblck** [comment]
Hi @maxandersen - Sorry for the delay, I've not had a large enough block of time to get started on this yet. I end up down a rabbit hole trying to figure out how Gradle works (I've never used this tool) and where to start (and give up or run out of time). Would you be able to give me a hint where to start?

**Comment by maxandersen** [comment]
Plugins can contribute these tasks. 

Spotless plugin (see near top) makes them available. 

So for grgit add it as plugin. 

Then look at "tag" task. That's more.or less just groovy.code that can do whatever it needs to do. 

In this case iterate over all files in the build install dir.
There is a property for it but I'm only on mobile atm :)

Hope that helps.

**Comment by instreest** [comment]
I looked into this. As already pointed out in this thread, `jbang wrapper install` copies the files with the `COPY_ATTRIBUTES` option, which carries over the timestamps of the source files.

## Why the source files are dated 1980 in the first place

The build deliberately turns off timestamp preservation for reproducibility:

```groovy
// build.gradle
// to enable reproducible builds
tasks.withType(AbstractArchiveTask) {
	preserveFileTimestamps = false
	reproducibleFileOrder = true
}
```

With `preserveFileTimestamps = false`, Gradle discards the real timestamps and writes a fixed constant into every archive entry. For ZIP that constant is `new GregorianCalendar(1980, 1, 1, 0, 0, 0)` in Gradle's `ZipEntryConstants` — months are 0-based, so 1980-02-01. Tar uses a different constant, 1970-01-02.

So this reproduces every time for a JBang installed from a release archive. A JBang built from source has normal timestamps, which is probably why it was hard to reproduce locally.

## Current behaviour, in one place

| File | Date today |
| --- | --- |
| `bin/jbang`, `bin/jbang.cmd`, `bin/jbang.ps1`, `bin/jbang.jar` inside the release archive | 1980-02-01 (zip) / 1970-01-02 (tar) |
| The same files once extracted/installed | unchanged, the entry timestamps are preserved |
| `jbang`, `jbang.cmd`, `jbang.ps1`, `.jbang/jbang.jar` produced by `jbang wrapper install` | inherited from the above |

One detail worth noting about the Homebrew listing posted above, where `bin/jbang` had a current date but everything else was dated 1980: that is not Homebrew fixing timestamps. The [formula](https://github.com/Homebrew/homebrew-core/blob/master/Formula/j/jbang.rb) runs

```ruby
inreplace "#{libexec}/bin/jbang", /^abs_jbang_dir=.*/, "abs_jbang_dir=#{libexec}/bin"
```

which rewrites that one file, so only `bin/jbang` gets a fresh modification time.

## Proposed fix

Fix it on the consuming side, in `Wrapper.java`, not in the build.

The 1980 constant in the archives is not a bug — it is exactly what `preserveFileTimestamps = false` is for — so the build side should stay as it is.

The wrapper files, on the other hand, are newly created files and should be dated accordingly. Keeping `COPY_ATTRIBUTES` and setting the modification time to the current time right after each copy does that, and leaves permissions unchanged.

## What this does not change

- The launcher scripts inside the `.zip` / `.tar` release archives keep their fixed 1980-02-01 (or 1970-01-02) entry timestamps.
- Consequently the scripts extracted by an installer keep those dates too.
- Only the modification time of the wrapper files is set. The creation time and the last access time are left as `Files.copy` leaves them, which on Windows can still be the archive's 1980 date. That keeps the change to a single line per copy, and the modification time is what file listings show.

---
_Investigated and drafted with [Claude Code](https://claude.ai/code). I have reviewed the findings myself before posting._
